jax/experimental/sparse/bcoo.py

```diff
@@ -610,6 +614,152 @@ def _bcoo_dot_general_abstract_eval(lhs_data, lhs_indices, rhs, *, dimension_num
+               [0, 3],
+               [1, 3],
+               [0, 0],
+               [0, 0],
+               [0, 0]], dtype=int32)
+
+  If the input data contains nonzeros at the corrected indices:
+  >>> data = jnp.arange(1, 8)
+  >>> indices = bcoo_mat.indices
+
+  After the out-of-bound indices correction:
+  >>> data, _ = _bcoo_correct_oob_indices_impl(data, indices, shape=(2, 5))
+
+  >>> data
+  DeviceArray([1, 2, 3, 4, 0, 0, 0], dtype=int32)
+  """
+  # TODO: support batch dimension?
+  props = _validate_bcoo(data, indices, shape)
+  if props.n_batch or props.n_dense:
+    raise NotImplementedError("`n_batch` and `n_dense` not yet supported.")
+  mask = (indices >= jnp.array(shape)).any(-1)
+  return jnp.where(mask, 0, data), jnp.where(mask[:, None], 0, indices)
+
+_bcoo_correct_oob_indices = xla.lower_fun(
+    _bcoo_correct_oob_indices_impl, multiple_results=True, new_style=True)
+
+_bcoo_dot_general_default_translation_rule = xla.lower_fun(
+    _bcoo_dot_general_impl, multiple_results=False, new_style=True)
+
+def _bcoo_dot_general_cuda_translation_rule(
+    ctx, avals_in, avals_out, lhs_data, lhs_indices, rhs, *, dimension_numbers,
+    lhs_shape):
+
+  c = ctx.builder
+
+  (lhs_contract, rhs_contract), _ = dimension_numbers
+  _, lhs_indices_aval, _, = avals_in
+  props = _validate_bcoo_indices(lhs_indices_aval, lhs_shape)
+
+  rhs_ndim = len(c.get_shape(rhs).dimensions())
+  if rhs_ndim == 1:
+    bcoo_dot_general_fn = cusparse.coo_matvec
+  elif rhs_ndim == 2:
+    bcoo_dot_general_fn = cusparse.coo_matmat
+    if rhs_contract[0] == 1:
+      rhs = xops.Transpose(rhs, permutation=[1, 0])
+
+  lhs_transpose = False
+  if props.n_sparse == 1:
+    # Converts lhs to a row vector.
+    col = xops.Collapse(lhs_indices, dimensions=[0, 1])
+    row = xops.Broadcast(xops.Constant(c, jnp.array(0, dtype=jnp.int32)),
+                         c.get_shape(col).dimensions())
+    lhs_shape = (1, lhs_shape[0])
+    dot_product = bcoo_dot_general_fn(
+        c, lhs_data, row, col, rhs, shape=lhs_shape, transpose=lhs_transpose)
+
+    if rhs_ndim == 1:
+      # Transforms a single-element array to a scalar.
+      return [xops.Reshape(dot_product, dimensions=[0], new_sizes=[])]
+    else:
+      return [xops.Collapse(dot_product, dimensions=[0, 1])]
+  elif props.n_sparse == 2:
+    row = xops.Collapse(
+        xops.Slice(lhs_indices,
+                   start_indices=[0, 0],
+                   limit_indices=[c.get_shape(lhs_indices).dimensions()[0], 1],
+                   strides=[1, 1]),
+        dimensions=[0, 1])
+    col = xops.Collapse(
+        xops.Slice(lhs_indices,
+                   start_indices=[0, 1],
+                   limit_indices=[c.get_shape(lhs_indices).dimensions()[0], 2],
+                   strides=[1, 1]),
+        dimensions=[0, 1])
+
+    if lhs_contract[0] == 0:
+      lhs_transpose = True
+
+    return [bcoo_dot_general_fn(
+        c, lhs_data, row, col, rhs, shape=lhs_shape, transpose=lhs_transpose)]
```

Comment:
The function implicitly returns `None` if `props.n_sparse` is not `1` or `2`, which would result in a huge & incomprehensible traceback through the JAX machinery. Another action-at-distance 😁

Reply:
Done.